src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -167,25 +164,141 @@ public static unsafe void MatMulX(bool add, AlignedArray mat, AlignedArray src,
-                        Vector256<float> x31 = Avx.LoadAlignedVector256(pMatTemp += ccol);
-                        Vector256<float> x02 = Avx.LoadAlignedVector256(pSrcCurrent);
-
-                        res0 = Avx.Add(res0, Avx.Multiply(x01, x02));
-                        res1 = Avx.Add(res1, Avx.Multiply(x11, x02));
-                        res2 = Avx.Add(res2, Avx.Multiply(x21, x02));
-                        res3 = Avx.Add(res3, Avx.Multiply(x31, x02));
-
-                        pSrcCurrent += 8;
-                        pMatCurrent += 8;
+                        int remainder = 0;
+                        if ((misalignment & 3) != 0)
+                        {
+                            // Handles cases where the data is not 32-bit aligned and we can't ever use aligned operations
+                            while (pSrcCurrent + 8 <= pSrcEnd)
+                            {
+                                float* pMatTemp = pMatCurrent;
+                                Vector256<float> x01 = Avx.LoadVector256(pMatTemp);
+                                Vector256<float> x11 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x21 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x31 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);
+
+                                res0 = Avx.Add(res0, Avx.Multiply(x01, vector));
+                                res1 = Avx.Add(res1, Avx.Multiply(x11, vector));
+                                res2 = Avx.Add(res2, Avx.Multiply(x21, vector));
+                                res3 = Avx.Add(res3, Avx.Multiply(x31, vector));
+
+                                pSrcCurrent += 8;
+                                pMatCurrent += 8;
+                            }
+                        }
+                        else
+                        {
+                            if (misalignment != 0)
+                            {
+                                // Handle cases where the data is not 256-bit aligned by doing an unaligned read and then
+                                // masking any elements that will be included in the first aligned read
+                                misalignment >>= 2;
+                                misalignment = 8 - misalignment;
+
+                                float* pMatTemp = pMatCurrent;
+                                Vector256<float> x01 = Avx.LoadVector256(pMatTemp);
+                                Vector256<float> x11 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x21 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x31 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);
```

Comment:
We should probably note that there is no attempt made to get both `pMat` and `pSrc` aligned. We only align `pMat` since it has significantly more reads.

Reply:
i will add a comment for that.